src/librustc/infer/error_reporting/mod.rs

```diff
@@ -485,12 +485,29 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         }
     }
 
-    fn note_error_origin(&self, err: &mut DiagnosticBuilder<'tcx>, cause: &ObligationCause<'tcx>) {
+    fn note_error_origin(
+        &self,
+        err: &mut DiagnosticBuilder<'tcx>,
+        cause: &ObligationCause<'tcx>,
+        exp_found: Option<ty::error::ExpectedFound<Ty<'tcx>>>,
+    ) {
         match cause.code {
             ObligationCauseCode::MatchExpressionArmPattern { span, ty } => {
                 if ty.is_suggestable() {  // don't show type `_`
                     err.span_label(span, format!("this match expression has type `{}`", ty));
                 }
+                if let Some(ty::error::ExpectedFound { found, .. }) = exp_found {
+                    if ty.is_box() && ty.boxed_ty() == found {
+                        if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
+                            err.span_suggestion_short_with_applicability(
```

Comment:
```suggestion
                            err.span_suggestion_with_applicability(
```

Although I dislike the resulting label inlined code suggestion, I think it is invaluable for newcomers that aren't using VSCode to _see_ the code, when crossing a certain line in the sand of jargon-heavy text, which I think "dereferencing" is on the other side of. Any programmer will _have_ to learn it quickly, but we want the experience to be good for _everyone_ without having to google too much.

Other than that nitpick (which you can disagree with), an enthusiastic r=me.